go/test/endtoend/orchestrator/orc_test.go

```diff
@@ -0,0 +1,361 @@
+		}
+		mysqlCtlProcessList = append(mysqlCtlProcessList, proc)
+	}
+
+	// Wait for mysql processes to start
+	for _, proc := range mysqlCtlProcessList {
+		if err := proc.Wait(); err != nil {
+			return nil, 1
+		}
+	}
+
+	return clusterInstance, 0
+
+}
+
+// Cases to test:
+// 1. create cluster with 1 replica and 1 rdonly, let orc choose master
+// verify rdonly is not elected, only replica
+// verify replication is setup
+func TestMasterElection(t *testing.T) {
+	defer cluster.PanicHandler(t)
+	clusterInstance, exitCode := createCluster(t, 1, 1)
+	if exitCode != 0 {
+		os.Exit(exitCode)
+	}
+	defer func() {
+		clusterInstance.Teardown()
+		tablets = []*cluster.Vttablet{}
+	}()
+	for _, tablet := range tablets {
+		// Create Database
+		err := tablet.VttabletProcess.CreateDB(keyspaceName)
+		require.NoError(t, err)
+
+		// Reset status, don't wait for the tablet status. We will check it later
+		tablet.VttabletProcess.ServingStatus = ""
+
+		// Start the tablet
+		err = tablet.VttabletProcess.Setup()
+		require.NoError(t, err)
+	}
+
+	defer func() {
+		// Kill tablets
+		killTablets(t)
+	}()
+
+	for _, tablet := range tablets {
+		err := tablet.VttabletProcess.WaitForTabletTypes([]string{"SERVING", "NOT_SERVING"})
+		require.NoError(t, err)
+	}
+
+	// Start orchestrator
+	clusterInstance.OrcProcess = clusterInstance.NewOrcProcess(path.Join(os.Getenv("PWD"), "test_config.json"))
+	err := clusterInstance.OrcProcess.Setup()
+	require.NoError(t, err)
+
+	//log.Exitf("error")
+	checkMasterTablet(t, tablets[0])
+}
+
+// 2. bring down master, let orc promote replica
+func TestDownMaster(t *testing.T) {
+	defer cluster.PanicHandler(t)
+	clusterInstance, exitCode := createCluster(t, 2, 0)
+	if exitCode != 0 {
+		os.Exit(exitCode)
+	}
+	defer func() {
+		clusterInstance.Teardown()
+		tablets = []*cluster.Vttablet{}
+	}()
+	for _, tablet := range tablets {
+		// Create Database
+		err := tablet.VttabletProcess.CreateDB(keyspaceName)
+		require.NoError(t, err)
+
+		// Reset status, don't wait for the tablet status. We will check it later
+		tablet.VttabletProcess.ServingStatus = ""
+
+		// Start the tablet
+		err = tablet.VttabletProcess.Setup()
+		require.NoError(t, err)
+	}
+
+	defer func() {
+		// Kill tablets
+		killTablets(t)
+	}()
+
+	for _, tablet := range tablets {
+		err := tablet.VttabletProcess.WaitForTabletTypes([]string{"SERVING", "NOT_SERVING"})
+		require.NoError(t, err)
+	}
+
+	// Start orchestrator
+	clusterInstance.OrcProcess = clusterInstance.NewOrcProcess(path.Join(os.Getenv("PWD"), "test_config.json"))
+	err := clusterInstance.OrcProcess.Setup()
+	require.NoError(t, err)
+
+	// find master from topo
+	curMaster := shardMasterTablet(t)
+	assert.NotNil(t, curMaster, "should have elected a master")
+
+	// Make the current master agent and database unavailable.
+	err = curMaster.VttabletProcess.TearDown()
+	require.NoError(t, err)
+	err = curMaster.MysqlctlProcess.Stop()
+	require.NoError(t, err)
+
+	// wait for some time for orc to fix up
```

Comment:
I'm just confused: where's the code that waits?

Reply:
later :) I'll remove the comment.